// wagmi-private/dist/wagmi-private.esm.js
import { _ as _defineProperty, C as Connector, n as normalizeChainId, a as _asyncToGenerator, b as ConnectorNotFoundError, U as UserRejectedRequestError, c as ChainNotConfiguredError, A as AddChainError, S as SwitchChainError } from './base-abbd9db4.esm.js';
export { A as AddChainError, c as ChainNotConfiguredError, C as Connector, b as ConnectorNotFoundError, S as SwitchChainError, U as UserRejectedRequestError, d as chain, e as defaultChains, f as defaultL2Chains, g as developmentChains, n as normalizeChainId } from './base-abbd9db4.esm.js';
import { Web3Provider } from '@ethersproject/providers';
import { getAddress, hexValue } from 'ethers/lib/utils';
import 'eventemitter3';

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = null != arguments[i] ? arguments[i] : {};
    i % 2 ? ownKeys(Object(source), !0).forEach(function (key) {
      _defineProperty(target, key, source[key]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) {
      Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
    });
  }

  return target;
}

var erc1155ABI = [{
  constant: true,
  inputs: [{
    internalType: 'address',
    name: '_owner',
    type: 'address'
  }, {
    internalType: 'uint256',
    name: '_id',
    type: 'uint256'
  }],
  name: 'balanceOf',
  outputs: [{
    internalType: 'uint256',
    name: '',
    type: 'uint256'
  }],
  payable: false,
  stateMutability: 'view',
  type: 'function'
}, {
  constant: true,
  inputs: [{
    internalType: 'uint256',
    name: '_id',
    type: 'uint256'
  }],
  name: 'uri',
  outputs: [{
    internalType: 'string',
    name: '',
    type: 'string'
  }],
  payable: false,
  stateMutability: 'view',
  type: 'function'
}];

var erc20ABI = [{
  constant: true,
  inputs: [],
  name: 'name',
  outputs: [{
    name: '',
    type: 'string'
  }],
  payable: false,
  stateMutability: 'view',
  type: 'function'
}, {
  constant: false,
  inputs: [{
    name: '_spender',
    type: 'address'
  }, {
    name: '_value',
    type: 'uint256'
  }],
  name: 'approve',
  outputs: [{
    name: '',
    type: 'bool'
  }],
  payable: false,
  stateMutability: 'nonpayable',
  type: 'function'
}, {
  constant: true,
  inputs: [],
  name: 'totalSupply',
  outputs: [{
    name: '',
    type: 'uint256'
  }],
  payable: false,
  stateMutability: 'view',
  type: 'function'
}, {
  constant: false,
  inputs: [{
    name: '_from',
    type: 'address'
  }, {
    name: '_to',
    type: 'address'
  }, {
    name: '_value',
    type: 'uint256'
  }],
  name: 'transferFrom',
  outputs: [{
    name: '',
    type: 'bool'
  }],
  payable: false,
  stateMutability: 'nonpayable',
  type: 'function'
}, {
  constant: true,
  inputs: [],
  name: 'decimals',
  outputs: [{
    name: '',
    type: 'uint8'
  }],
  payable: false,
  stateMutability: 'view',
  type: 'function'
}, {
  constant: true,
  inputs: [{
    name: '_owner',
    type: 'address'
  }],
  name: 'balanceOf',
  outputs: [{
    name: 'balance',
    type: 'uint256'
  }],
  payable: false,
  stateMutability: 'view',
  type: 'function'
}, {
  constant: true,
  inputs: [],
  name: 'symbol',
  outputs: [{
    name: '',
    type: 'string'
  }],
  payable: false,
  stateMutability: 'view',
  type: 'function'
}, {
  constant: false,
  inputs: [{
    name: '_to',
    type: 'address'
  }, {
    name: '_value',
    type: 'uint256'
  }],
  name: 'transfer',
  outputs: [{
    name: '',
    type: 'bool'
  }],
  payable: false,
  stateMutability: 'nonpayable',
  type: 'function'
}, {
  constant: true,
  inputs: [{
    name: '_owner',
    type: 'address'
  }, {
    name: '_spender',
    type: 'address'
  }],
  name: 'allowance',
  outputs: [{
    name: '',
    type: 'uint256'
  }],
  payable: false,
  stateMutability: 'view',
  type: 'function'
}, {
  payable: true,
  stateMutability: 'payable',
  type: 'fallback'
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    name: 'owner',
    type: 'address'
  }, {
    indexed: true,
    name: 'spender',
    type: 'address'
  }, {
    indexed: false,
    name: 'value',
    type: 'uint256'
  }],
  name: 'Approval',
  type: 'event'
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    name: 'from',
    type: 'address'
  }, {
    indexed: true,
    name: 'to',
    type: 'address'
  }, {
    indexed: false,
    name: 'value',
    type: 'uint256'
  }],
  name: 'Transfer',
  type: 'event'
}];

var erc721ABI = [{
  inputs: [{
    internalType: 'uint256',
    name: 'tokenId',
    type: 'uint256'
  }],
  name: 'ownerOf',
  outputs: [{
    internalType: 'address',
    name: '',
    type: 'address'
  }],
  stateMutability: 'view',
  type: 'function'
}, {
  inputs: [{
    internalType: 'uint256',
    name: 'tokenId',
    type: 'uint256'
  }],
  name: 'tokenURI',
  outputs: [{
    internalType: 'string',
    name: '',
    type: 'string'
  }],
  stateMutability: 'view',
  type: 'function'
}];

// https://github.com/ethers-io/ethers.js/blob/master/packages/units/src.ts/index.ts#L10-L18
var units = ['wei', 'kwei', 'mwei', 'gwei', 'szabo', 'finney', 'ether'];

class InjectedConnector extends Connector {
  constructor(config) {
    super(_objectSpread2(_objectSpread2({}, config), {}, {
      options: undefined
    }));

    _defineProperty(this, "id", 'injected');

    _defineProperty(this, "name", void 0);

    _defineProperty(this, "ready", typeof window != 'undefined' && !!window.ethereum);

    _defineProperty(this, "_provider", void 0);

    _defineProperty(this, "onAccountsChanged", accounts => {
      if (accounts.length === 0) this.emit('disconnect');else this.emit('change', {
        account: accounts[0]
      });
    });

    _defineProperty(this, "onChainChanged", chainId => {
      var id = normalizeChainId(chainId);
      var unsupported = this.isChainUnsupported(id);
      this.emit('change', {
        chain: {
          id,
          unsupported
        }
      });
    });

    _defineProperty(this, "onDisconnect", () => {
      this.emit('disconnect');
    });

    var name = 'Injected';

    if (typeof window !== 'undefined' && window.ethereum) {
      if (window.ethereum.isMetaMask) name = 'MetaMask';
      if (window.ethereum.isCoinbaseWallet) name = 'Coinbase Wallet';
    }

    this.name = name;
  }

  connect() {
    var _this = this;

    return _asyncToGenerator(function* () {
      try {
        var provider = _this.getProvider();

        if (!provider) throw new ConnectorNotFoundError();

        if (provider.on) {
          provider.on('accountsChanged', _this.onAccountsChanged);
          provider.on('chainChanged', _this.onChainChanged);
          provider.on('disconnect', _this.onDisconnect);
        }

        var account = yield _this.getAccount();
        var id = yield _this.getChainId();

        var unsupported = _this.isChainUnsupported(id);

        return {
          account,
          chain: {
            id,
            unsupported
          },
          provider
        };
      } catch (error) {
        if (error.code === 4001) throw new UserRejectedRequestError();
        throw error;
      }
    })();
  }

  disconnect() {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      var provider = _this2.getProvider();

      if (!(provider !== null && provider !== void 0 && provider.removeListener)) return;
      provider.removeListener('accountsChanged', _this2.onAccountsChanged);
      provider.removeListener('chainChanged', _this2.onChainChanged);
      provider.removeListener('disconnect', _this2.onDisconnect);
    })();
  }

  getAccount() {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      var provider = _this3.getProvider();

      if (!provider) throw new ConnectorNotFoundError();
      var accounts = yield provider.request({
        method: 'eth_requestAccounts'
      }); // return checksum address

      return getAddress(accounts[0]);
    })();
  }

  getChainId() {
    var _this4 = this;

    return _asyncToGenerator(function* () {
      var provider = _this4.getProvider();

      if (!provider) throw new ConnectorNotFoundError();
      return yield provider.request({
        method: 'eth_chainId'
      }).then(normalizeChainId);
    })();
  }

  getProvider() {
    if (typeof window !== 'undefined' && !!window.ethereum) this._provider = window.ethereum;
    return this._provider;
  }

  getSigner() {
    var _this5 = this;

    return _asyncToGenerator(function* () {
      var provider = _this5.getProvider();

      var account = yield _this5.getAccount();
      return new Web3Provider(provider).getSigner(account);
    })();
  }

  isAuthorized() {
    var _this6 = this;

    return _asyncToGenerator(function* () {
      try {
        var provider = _this6.getProvider();

        if (!provider) throw new ConnectorNotFoundError();
        var accounts = yield provider.request({
          method: 'eth_accounts'
        });
        var account = accounts[0];
        return !!account;
      } catch (_unused) {
        return false;
      }
    })();
  }

  switchChain(chainId) {
    var _this7 = this;

    return _asyncToGenerator(function* () {
      var provider = _this7.getProvider();

      if (!provider) throw new ConnectorNotFoundError();
      var id = hexValue(chainId);

      try {
        yield provider.request({
          method: 'wallet_switchEthereumChain',
          params: [{
            chainId: id
          }]
        });
      } catch (error) {
        // Indicates chain is not added to MetaMask
        if (error.code === 4902) {
          try {
            var _chain$blockExplorers;

            var chain = _this7.chains.find(x => x.id === chainId);

            if (!chain) throw new ChainNotConfiguredError();
            yield provider.request({
              method: 'wallet_addEthereumChain',
              params: [{
                chainId: id,
                chainName: chain.name,
                nativeCurrency: chain.nativeCurrency,
                rpcUrls: chain.rpcUrls,
                blockExplorerUrls: (_chain$blockExplorers = chain.blockExplorers) === null || _chain$blockExplorers === void 0 ? void 0 : _chain$blockExplorers.map(x => x.url)
              }]
            });
          } catch (addError) {
            throw new AddChainError();
          }
        } else if (error.code === 4001) throw new UserRejectedRequestError();else throw new SwitchChainError();
      }
    })();
  }

  watchAsset(_ref) {
    var _this8 = this;

    return _asyncToGenerator(function* () {
      var {
        address,
        decimals = 18,
        image,
        symbol
      } = _ref;

      var provider = _this8.getProvider();

      if (!provider) throw new ConnectorNotFoundError();
      yield provider.request({
        method: 'wallet_watchAsset',
        params: {
          type: 'ERC20',
          options: {
            address,
            decimals,
            image,
            symbol
          }
        }
      });
    })();
  }

}

export { InjectedConnector, erc1155ABI, erc20ABI, erc721ABI, units };
