// wagmi-private/connectors/walletConnect/dist/wagmi-private-connectors-walletConnect.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var base = require('../../../dist/base-900b7d10.cjs.dev.js');
var providers = require('@ethersproject/providers');
var WalletConnectProvider = require('@walletconnect/ethereum-provider');
var utils = require('ethers/lib/utils');
require('eventemitter3');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var WalletConnectProvider__default = /*#__PURE__*/_interopDefault(WalletConnectProvider);

class WalletConnectConnector extends base.Connector {
  constructor(config) {
    super(config);

    base._defineProperty(this, "id", 'walletConnect');

    base._defineProperty(this, "name", 'WalletConnect');

    base._defineProperty(this, "ready", true);

    base._defineProperty(this, "_provider", void 0);

    base._defineProperty(this, "onAccountsChanged", accounts => {
      if (accounts.length === 0) this.emit('disconnect');else this.emit('change', {
        account: accounts[0]
      });
    });

    base._defineProperty(this, "onChainChanged", chainId => {
      var id = base.normalizeChainId(chainId);
      var unsupported = this.isChainUnsupported(id);
      this.emit('change', {
        chain: {
          id,
          unsupported
        }
      });
    });

    base._defineProperty(this, "onDisconnect", () => {
      this.emit('disconnect');
    });
  }

  connect() {
    var _this = this;

    return base._asyncToGenerator(function* () {
      try {
        var provider = _this.getProvider(true);

        provider.on('accountsChanged', _this.onAccountsChanged);
        provider.on('chainChanged', _this.onChainChanged);
        provider.on('disconnect', _this.onDisconnect);
        var accounts = yield provider.enable();
        var account = utils.getAddress(accounts[0]);
        var id = yield _this.getChainId();

        var unsupported = _this.isChainUnsupported(id);

        return {
          account,
          chain: {
            id,
            unsupported
          },
          provider: new providers.Web3Provider(provider)
        };
      } catch (error) {
        if (error.message === 'User closed modal') throw new base.UserRejectedRequestError();
        throw error;
      }
    })();
  }

  disconnect() {
    var _this2 = this;

    return base._asyncToGenerator(function* () {
      var provider = _this2.getProvider();

      yield provider.disconnect();
      provider.removeListener('accountsChanged', _this2.onAccountsChanged);
      provider.removeListener('chainChanged', _this2.onChainChanged);
      provider.removeListener('disconnect', _this2.onDisconnect);
      typeof localStorage !== 'undefined' && localStorage.removeItem('walletconnect');
    })();
  }

  getAccount() {
    var _this3 = this;

    return base._asyncToGenerator(function* () {
      var provider = _this3.getProvider();

      var accounts = provider.accounts; // return checksum address

      return utils.getAddress(accounts[0]);
    })();
  }

  getChainId() {
    var _this4 = this;

    return base._asyncToGenerator(function* () {
      var provider = _this4.getProvider();

      var chainId = base.normalizeChainId(provider.chainId);
      return chainId;
    })();
  }

  getProvider(create) {
    if (!this._provider || create) this._provider = new WalletConnectProvider__default["default"](this.options);
    return this._provider;
  }

  getSigner() {
    var _this5 = this;

    return base._asyncToGenerator(function* () {
      var provider = _this5.getProvider();

      var account = yield _this5.getAccount();
      return new providers.Web3Provider(provider).getSigner(account);
    })();
  }

  isAuthorized() {
    var _this6 = this;

    return base._asyncToGenerator(function* () {
      try {
        var account = yield _this6.getAccount();
        return !!account;
      } catch (_unused) {
        return false;
      }
    })();
  }

}

exports.WalletConnectConnector = WalletConnectConnector;
